File: 2020/day07/aoc.cpp
#include "task_runner.h"

struct Task {
    int start;
    std::unordered_map<std::string, int> node2id;
    std::unordered_map<int, std::string> id2node;

    static constexpr auto NMAX = 100'000;
    std::vector<std::pair<int, int>> g[NMAX];
    std::vector<std::pair<int, int>> g_t[NMAX];

    void read() {
        int cnt = 0;
        auto get_node = [&](const auto& n) {
            if (node2id.find(n) != node2id.end()) {
                return node2id[n];
            }

            int id = ++cnt;
            node2id[n] = id;
            id2node[id] = n;
            return id;
        };

        start = get_node("shiny gold");

        std::string line;
        while (std::getline(std::cin, line)) {
            if (!line.empty()) {
                if (line.find("no other bags.") != line.npos) {
                    continue;
                }

                std::istringstream iss{line};

                std::string base_color1, base_color2;
                std::string bags, contain;
                iss >> base_color1 >> base_color2 >> bags >> contain;
                auto b = get_node(fmt::format("{} {}", base_color1, base_color2));

                int c;
                std::string dep_color1, dep_color2;
                while (iss >> c >> dep_color1 >> dep_color2 >> bags) {
                    auto d = get_node(fmt::format("{} {}", dep_color1, dep_color2));

                    g[b].emplace_back(c, d);
                    g_t[d].emplace_back(c, b);
                }
            }
        }
    }

    auto dfs(auto node, const auto& g, auto& visited, auto& cache) {
        if (visited[node]) {
            return std::pair{0, cache[node].second};
        }
        visited[node] = 1;

        return cache[node] = std::ranges::accumulate(g[node], std::pair{1, 0}, [&](auto sum, const auto& neigbour) {
            auto [c, id] = neigbour;
            auto [c1, c2] = dfs(id, g, visited, cache);
            return std::pair{
                sum.first + c1,
                sum.second + c * (1 + c2),
            };
        });
    }

    auto compute() {
        std::vector<bool> visited(NMAX, false);
        std::vector<bool> visited_t(NMAX, false);
        std::unordered_map<int, std::pair<int, int>> cache, cache_t;

        return std::pair{
            dfs(start, g_t, visited_t, cache_t).first - 1,
            dfs(start, g, visited, cache).second,
        };
    }

    void print(const auto& result) {
        std::cout << result.first << "\n";
        std::cout << result.second << "\n";
    }

    auto solve() {
        read();
        print(compute());
    }
};

int main() {
    TaskRunner<Task>::create()->set_stream("in", std::cin).solve();
    // TaskRunner<Task>::create()->set_stream("in", std::cin).set_stream("out", std::cout).solve();
    return 0;
}

File: 2020/day07/aoc.py
import math
import sys


class Task:
    def read(self, file='in'):
        return [line.strip() for line in open(file).read().strip().split('\n')]

    def build_graph(self, input):
        is_dep = {}
        has_dep = {}

        for line in input:
            base_color = line.split('bags contain')[0].strip()

            deps = line.split('contain')[1].strip(' .')
            if deps == 'no other bags':
                continue

            for dep in deps.split(','):
                count = int(dep.strip().split(' ')[0])
                dep_color = ' '.join(dep.strip().split(' ')[1:3])

                has_dep.setdefault(base_color, []).append((count, dep_color))
                is_dep.setdefault(dep_color, []).append((count, base_color))

        return (is_dep, has_dep)

    def compute(self, input):
        def dfs(node, is_dep, visited):
            if node in visited:
                return 0
            visited[node] = True

            return 1 + sum(map(lambda cx: dfs(cx[1], is_dep, visited), is_dep.get(node, [])))

        def dfs_transpose(node, has_dep, visited, cache):
            if node in visited:
                return cache[node]
            visited[node] = True

            cache[node] = sum(map(lambda cx: cx[0] + cx[0] * dfs_transpose(cx[1],
                                                                           has_dep, visited, cache), has_dep.get(node, [])))
            return cache[node]

        start = 'shiny gold'
        is_dep, has_dep = self.build_graph(input)

        cnt_nodes = dfs(start, is_dep, {}) - 1  # don't count start
        cnt_deps = dfs_transpose(start, has_dep, {}, {})

        return (cnt_nodes, cnt_deps)

    def print(self, result):
        print('{}\n{}\n'.format(result[0], result[1]))

    def solve(self):
        input = self.read()
        self.print(self.compute(input))


Task().solve()
